---
title: "Data Access: Downloading SRA and FastQ data in R using SRAdb fasp/ftp protocols"
author: "Beatriz Manso"
date: '2022-02-23'
output: 
  distill::distill_article:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
    code_folding: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

DNA sequences of short reads generated by high throughput sequencing are stored in the Sequence Read Archive (SRA), a bioinformatics database.

SRA is an inter-institutional collaboration among NCBI SRA, EBI ENA and DDBJ DRA.

Using SRA, organisms are arranged hierarchically based on sequence and are classified into six categories:

-   SR**A** = Submission; a virtual container for other objects (below).
-   SR**P** = Study/**P**roject; describes project's metadata.
-   SR**S** = **S**ample; describes the sample which was sequenced.
-   SR**X** E**x**periment; describes the library, platform, and processing parameters
-   SR**R** = **R**un; contains sequencing files
-   SR**Z** = Analy**z**is; contains BAM files


# Methods

## 1. Set Working directory

```{r}
setwd("C:/Users/manso/OneDrive - University of West London/MSc Bioinformatics - UWL/7.MSc Bioinformatics Project/W2 - Sample size and power analysis/practical")
```


## 2. Instaling necessary packages

```{r}
if (!requireNamespace("BiocMansger", quietly=TRUE))
  install.packages("BiocManager", repos ="http://cran.us.r-project.org")
BiocManager::install("SRAdb")

```


## 3. Load libraries

```{r}
library(SRAdb)
```


## 4. Download the SRAmetadb sqlite database

This code chunk will only download if the file does not exist. It can take a while to download(~2.4GB zip file)

```{r}
if( ! file.exists('SRAmetadb.sqlite.gz') ) {
  sra_dbname <- getSRAdbFile()
  } else { 
  sra_dbname <- 'SRAmetadb.sqlite.gz'
}
sra_dbname <- file.path(system.file('extdata', package='SRAdb'), 'SRAmetadb_demo.sqlite')

#Connect to the database
sra_con <- dbConnect( dbDriver("SQLite"), sra_dbname )
```


## 5. List fasp addresses of sra files associated with "SRR" ID

```{r}
listSRAfile( c("SRR000648","SRR000657"), sra_con, fileType = 'sra', srcType='fasp')
```


## 6. Download SRA files using fasp protocol

```{r}
ascpCMD <- 'ascp -QT -l 300m -i /usr/local/aspera/connect/etc/asperaweb_id_dsa.putty'

SRAFiles <- getSRAfile( c("SRR000648","SRR000657"), sra_con, 
                        fileType = 'sra', srcType = 'fasp', ascpCMD = ascpCMD )
```

-T : disable encryption for maximum throughput -l : set the target transfer rate in kbps -i : to specify a private key file.


## 7. Get information about the fastQ files for sample accession IDs

```{r}
getFASTQinfo( in_acc = c("SRR000648","SRR000657"), sra_con, srcType = 'fasp' )

```


## 8. Download FASTQ files using ftp protocol

```{r}
getFASTQfile( in_acc = c("SRR000648","SRR000657"), sra_con, destDir = getwd(),
              srcType = 'ftp', makeDirectory = FALSE, method = 'curl', ascpCMD = NULL)
```

The sra or sra-lite data files are downloaded from NCBI SRA and the fastq files are downloaded from EBI ENA.

The files are donloaded in to the working directory.


## 9. Finaly, it is recommended to disconnect from the database

```{r}
dbDisconnect( sra_con )
```
